c7n/tags.py

```diff
@@ -917,6 +919,125 @@ def process_resource_set(self, resource_set, tags):
+    )
+
+    def get_permissions(self):
+        return self.manager.action_registry.get('tag').permissions
+
+    def validate(self):
+        related_resource = self.data['resource']
+        if related_resource not in all_resources.keys():
+            raise PolicyValidationError(
+                "Error: Invalid resource type selected: %s" % related_resource
+            )
+        # ideally should never raise here since we shouldn't be applying this
+        # action to a resource if it doesn't have a tag action implemented
+        if self.manager.action_registry.get('tag') is None:
+            raise PolicyValidationError(
+                "Error: Tag action missing on resource"
+            )
+        return self
+
+    def process(self, resources):
+        tag_keys = self.data['copy_tags']
+        related_key = self.data['key']
+        related_type = self.data['resource']
+
+        related_resource_tuple = list(
+            zip(jmespath.search('[].%s' % related_key, resources), resources))
+
+        related_ids = [r[0] for r in related_resource_tuple]
+        related_tag_map = self.get_resource_tag_map(related_type, related_ids)
+        missing_related_tags = list(set(related_ids).difference(related_tag_map.keys()))
+
+        if not self.data.get('skip_missing', True) and missing_related_tags:
+            raise PolicyExecutionError(
+                "Unable to find all %s resources associated with %s" % (related_type, related_ids))
+
+        for related, r in related_resource_tuple:
+            if related in missing_related_tags or not related_tag_map[related]:
+                self.log.warning(
+                    'Tags not found for related resource: %s, skipping %s' % (related, r))
+                continue
+            related_tags = self.format_tags(related_tag_map[related])
+            self.process_resource(r, related_tags, tag_keys)
+
+    def process_resource(self, r, related_tags, tag_keys):
+        add_tags = {}
+        if '*' in tag_keys:
+            add_tags = related_tags
+        else:
+            for t in tag_keys:
+                if t in related_tags:
+                    add_tags[t] = related_tags[t]
+        if not add_tags:
+            return
+        # rely on resource manager tag action implementation as it can differ between resources
+        data = {'type': 'tag', 'tags': add_tags}
+        tag_action = self.manager.action_registry.get('tag')(data, self.manager)
+        self.log.info('Adding related tags: %s to resource: %s' % (add_tags, r))
+        tag_action.process(resources=[r])
+
+    def get_resource_tag_map(self, r_type, ids):
+        """
+        Returns a mapping of {resource_id: resource['Tags']}
+        """
+        manager = self.manager.get_resource_manager(r_type)
+        resources = manager.get_resources([])
+        r_id = manager.resource_type.id
+        return {r[r_id]: r.get('Tags', []) for r in resources if r[r_id] in ids}
+
+    def format_tags(self, tags):
```

Comment:
one line functions that are used in one place aren't particularly useful vs inlining the one line.